server/image_list.go

```diff
@@ -33,16 +33,17 @@ func ConvertImage(from *storage.ImageResult) *types.Image {
 		return nil
 	}
 
-	repoTags := []string{"<none>:<none>"}
+	repoTags := []string{}
+	repoDigests := []string{}
+
 	if len(from.RepoTags) > 0 {
 		repoTags = from.RepoTags
-	} else if from.PreviousName != "" {
-		repoTags = []string{from.PreviousName + ":<none>"}
+	} else if from.PreviousName != "" && from.Digest != "" {
+		repoDigests = []string{from.PreviousName + "@" + string(from.Digest)}
```

Comment:
If we build a container image with the same tag two times in a row locally, then the first built image has:
- no repo tags set
- no repo digests set
- a previous name set
- a Digest set
Using the previous name and the `from.Digest` can then be used to assemble a single repoDigest together.

WDYT @mtrmac ?

Reply:
Intuitively this looks good — but it’s just a guess, ultimately I don’t think we can avoid tracing down all the users, collecting the full set of constraints, and documenting what we are doing and how it fits together (and how can future users work with the API as intended). It’s the only way to escape the “here’s a test case that breaks and here’s a PR that fixes that one test case” treadmill.

- Does a built image _always_ have a name with this logic? (A plain `buildah commit` just returns an ID IIRC.) If not, isn’t there a caller that needs updating to handle that? (https://github.com/cri-o/cri-o/issues/2454 , or maybe the crictl code?)
-  (For me personally, this is a bit hard to follow.
   ```go
   repoTags = from.RepoTags // or just inline this into the `types.Image{}` constructor
   
   if len(from.RepoTags) == 0 && from.PreviousName… {
      repoDigests = []string{}
   }
   if len(from.RepoDigests) > 0 {
     repoDigests = append(repoDigests, from.RepoDigests...)
   }
   ```
   would feel more natural, one “paragraph” per goal.
- Actually, should we be using the PreviousName value if RepoDigests is set at all?

Reply:
Thank you, I simplified the `if` `else` paths and moved the `PreviousName` part to the actual digest handling. I added documentation of all necessary use cases and linked it from the corresponding sections.